src/realm/sync/noinst/client_history_impl.cpp

```diff
@@ -359,128 +359,161 @@ void ClientHistory::integrate_server_changesets(const SyncProgress& progress,
+        VersionID old_version;
+        TransactionRef transact = m_db->start_write(); // Throws
+        old_version = transact->get_version_of_current_transaction();
+        version_type local_version = old_version.version;
+
+        ensure_updated(local_version); // Throws
+        prepare_for_write();           // Throws
+
+        REALM_ASSERT(transact->get_sync_file_id() != 0);
+
+        std::uint_fast64_t downloaded_bytes_in_subset = 0;
+
+        try {
+            // Transform and apply one subset of changesets at a time.
+            while (changeset_ndx < num_changesets) {
+                REALM_ASSERT(subsets_to_integrate > 0);
+
+                auto changeset_ndx_end = changeset_ndx;
+                std::uint_fast64_t instructions_so_far_count = 0;
+                // Find the changesets that are part of the current subset.
+                do {
+                    const RemoteChangeset& changeset = incoming_changesets[changeset_ndx_end];
+                    REALM_ASSERT(changeset.last_integrated_local_version <= local_version);
+                    REALM_ASSERT(changeset.origin_file_ident > 0 &&
+                                 changeset.origin_file_ident != transact->get_sync_file_id());
+                    downloaded_bytes_in_subset += changeset.original_changeset_size;
+
+                    // It is possible that the synchronization history has been trimmed
+                    // to a point where a prefix of the merge window is no longer
+                    // available, but this can only happen if that prefix consisted
+                    // entirely of upload skippable entries. Since such entries (those
+                    // that are empty or of remote origin) will be skipped by the
+                    // transformer anyway, we can simply clamp the beginning of the
+                    // merge window to the beginning of the synchronization history,
+                    // when this situation occurs.
+                    //
+                    // See trim_sync_history() for further details.
+                    if (changesets[changeset_ndx_end].last_integrated_remote_version < m_sync_history_base_version)
+                        changesets[changeset_ndx_end].last_integrated_remote_version = m_sync_history_base_version;
+
+                    instructions_so_far_count += changesets[changeset_ndx_end].size();
+                    ++changeset_ndx_end;
+                } while (changeset_ndx_end < num_changesets && instructions_so_far_count < instructions_per_subset);
+
+                Transformer& transformer = get_transformer(); // Throws
+                Transformer::Reporter* reporter = nullptr;
+                transformer.transform_remote_changesets(
+                    *this, transact->get_sync_file_id(), local_version, changesets.data() + changeset_ndx,
+                    changeset_ndx_end - changeset_ndx, reporter, &logger); // Throws
+
+                while (changeset_ndx < changeset_ndx_end) {
```

Comment:
I think you can just do `for (;changeset_ndx < changeset_ndx_end; ++changeset_ndx) { ... }` here